python/spark_sklearn/keyed_models.py

```diff
@@ -0,0 +1,407 @@
+from pyspark.sql.functions import explode, udf
+from pyspark.sql.types import *
+from pyspark.sql.types import _type_mappings, UserDefinedType, NumericType
+
+from spark_sklearn.group_apply import gapply
+
+class _SparkSklearnEstimatorUDT(UserDefinedType):
+    """
+    SQL user-defined type (UDT) for scikit-learn estimator
+    """
+
+    @classmethod
+    def sqlType(cls):
+        return StringType() if sys.version_info[0] < 3 else BinaryType()
+
+    @classmethod
+    def module(cls):
+        return "spark_sklearn.keyed_models"
+
+    def serialize(self, obj):
+        return pickle.dumps(obj.estimator)
+
+    def deserialize(self, datum):
+        return SparkSklearnEstimator(pickle.loads(datum))
+
+    def simpleString(self):
+        return "sklearn-estimator"
+
+class SparkSklearnEstimator(object):
+    """:class:`SparkSklearnEstimator` is a wrapper for containing scikit-learn estimators in
+    dataframes - any estimators need to be stored inside the wrapper class to be properly
+    serialized/deserialized in dataframe operations.
+    """
+
+    __UDT__ = _SparkSklearnEstimatorUDT()
+
+    def __init__(self, estimator):
+        """Initializes with the parameter estimator.
+
+        :param: estimator: scikit-learn estimator to contain.
+        """
+        self._estimator = estimator
+
+    @property
+    def estimator(self):
+        return self._estimator
+
+    def __getattr__(self, item):
+        if hasattr(self._estimator, item):
+            return getattr(self._estimator, item)
+        raise AttributeError()
+
+def _isOneDimensional(schema, xCol):
+    xType = schema[xCol].dataType
+    return isinstance(xType, NumericType)
+
+def _validateXCol(schema, xCol):
+    if not _isOneDimensional(schema, xCol) and schema[xCol].dataType != Vector.__UDT__:
+        raise TypeError("Input column {} is neither a numeric type nor a vector".format(xCol))
+
+def _prepareXCol(series, is1D):
+    if is1D:
+        return series.values.reshape(-1, 1)
+    else:
+        return np.vstack(series.apply(lambda v: v.toArray()))
+
+@inherit_doc
+class KeyedEstimator(pyspark.ml.Estimator):
+    """A :class:`KeyedEstimator` provides an interface for training per-key scikit-learn estimators.
+
+    The :class:`KeyedEstimator` can be part of any Spark ML pipeline provided the columns are
+    appropriately matched.
+
+    Currently, the class provides a generalization for scikit-learn transformers and predictors
+    (per-key clustering is currently unsupported). Because these scikit-learn estimators
+    all derive from the same base type (yielding the same API), yet have different expectations
+    for what methods should be called and with what arguments, this class enumerates two different
+    types of behavior:
+
+    1. `"transformer"`
+
+        Examples: `sklearn.decomposition.PCA`, `sklearn.cluster.KMeans`
+
+        In this case, the estimator will aggregate the all input features for a given key into a
```

Comment:
RESPONSE REQUESTED: vectors not necessary if 1D. I'll leave comment as-is unless you tell me otherwise.